src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs

```diff
@@ -0,0 +1,173 @@
+            Requires.NotNull(activeConfiguredProjectSubscriptionService, nameof(activeConfiguredProjectSubscriptionService));
+            Requires.NotNull(launchProviders, nameof(launchProviders));
+
+            _projectVsServices = projectVsServices;
+            _startupProjectsListService = serviceProvider.GetService<IVsStartupProjectsListService, SVsStartupProjectsListService>();
+            _threadingService = threadingService;
+            _activeConfiguredProjectSubscriptionService = activeConfiguredProjectSubscriptionService;
+            _launchProviders = launchProviders;
+
+            WrapperMethodCaller = new DataFlowExtensionMethodCaller(new DataFlowExtensionMethodWrapper());
+        }
+
+        [ProjectAutoLoad]
+        [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+        internal async Tasks.Task OnProjectFactoryCompletedAsync()
+        {
+            ConfigurationGeneral projectProperties =
+                await _projectVsServices.ActiveConfiguredProjectProperties.GetConfigurationGeneralPropertiesAsync().ConfigureAwait(false);
+            _guid = new Guid((string)await projectProperties.ProjectGuid.GetValueAsync().ConfigureAwait(false));
+            Assumes.False(_guid == Guid.Empty);
+
+            await InitializeAsync().ConfigureAwait(false);
+        }
+
+        public async Tasks.Task InitializeAsync()
+        {
+            await AddOrRemoveProjectFromStartupProjectList(initialize: true).ConfigureAwait(false);
+
+            var watchedEvaluationRules = Empty.OrdinalIgnoreCaseStringSet.Add(ConfigurationGeneral.SchemaName);
+            var evaluationBlock = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(
+                ConfigurationGeneralRuleBlock_ChangedAsync);
+            IReceivableSourceBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>> sourceBlock = 
+                _activeConfiguredProjectSubscriptionService.ProjectRuleSource.SourceBlock;
+            _evaluationSubscriptionLink = WrapperMethodCaller.LinkTo(
+                sourceBlock,
+                evaluationBlock,
+                ruleNames: watchedEvaluationRules, suppressVersionOnlyUpdates:true
+                );
+        }
+
+        internal async Tasks.Task ConfigurationGeneralRuleBlock_ChangedAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> e)
+        {
+            IProjectChangeDescription projectChange = e.Value.ProjectChanges[ConfigurationGeneral.SchemaName];
+
+            /* Currently  we watch for the change in the OutputType to check if a project is debuggable.
+               There are other cases where the OutputType will remain the same and still the ability to debuggable a project could change
+               For eg: A project's OutputType could be a Lib and an execution entry point could be added or removed
+               */
+            if (projectChange.Difference.ChangedProperties.Contains(ConfigurationGeneral.OutputTypeProperty))
```

Comment:
Please file a bug around this. We should make this correct - anything that affects when a debugger can launch should be factored in this (such as whether I've set a custom launch exe or browser, etc).